src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AddClassProjectCommand.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.Input;
+using Microsoft.VisualStudio.ProjectSystem.Input;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using System;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands
+{
+    [ProjectCommand(CommandGroup.VisualStudioStandard97, VisualStudioStandard97CommandId.AddClass)]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class AddClassProjectCommand : AbstractSingleNodeProjectCommand
+    {
+        private readonly IPhysicalProjectTree _projectTree;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly SVsServiceProvider _serviceProvider;
+
+        [ImportingConstructor]
+        public AddClassProjectCommand(IPhysicalProjectTree projectTree, IUnconfiguredProjectVsServices projectVsServices, SVsServiceProvider serviceProvider)
+        {
+            Requires.NotNull(projectTree, nameof(IPhysicalProjectTree));
+            Requires.NotNull(projectVsServices, nameof(IUnconfiguredProjectVsServices));
+            Requires.NotNull(serviceProvider, nameof(SVsServiceProvider));
+
+            _projectTree = projectTree;
+            _projectVsServices = projectVsServices;
+            _serviceProvider = serviceProvider;
+        }
+
+        protected override Task<CommandStatusResult> GetCommandStatusAsync(IProjectTree node, bool focused, string commandText, CommandStatus progressiveStatus)
+        {
+            if (_projectTree.NodeCanHaveAdditions(node))
+            {
+                return GetCommandStatusResult.Handled(commandText, CommandStatus.Enabled);
+            }
+            else
+            {
+                return GetCommandStatusResult.Unhandled;
+            }
+        }
+
+        protected override async Task<bool> TryHandleCommandAsync(IProjectTree node, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut)
+        {
+            if (!_projectTree.NodeCanHaveAdditions(node))
+            {
+                return false;
+            }
+
+
+            string strBrowseLocations = _projectTree.TreeProvider.GetAddNewItemDirectory(node);
+            IVsAddProjectItemDlg addItemDialog = _serviceProvider.GetService<IVsAddProjectItemDlg, SVsAddProjectItemDlg>();
+            Assumes.Present(addItemDialog);
+
+            string strFilter = string.Empty;
+            int iDontShowAgain;
```

Comment:
Nit: Move some of the inputs/outputs closer to the calling of the method itself - so that when code gets moved around, so do these.